storage/reader.go

```diff
@@ -77,9 +80,76 @@ type ReaderObjectAttrs struct {
 //
 // The caller must call Close on the returned Reader when done reading.
 func (o *ObjectHandle) NewReader(ctx context.Context) (*Reader, error) {
+	if o.c.gc != nil {
+		return o.newRangeReaderWithGRPC(ctx, 0, 0)
+	}
 	return o.NewRangeReader(ctx, 0, -1)
 }
 
+// newRangeReaderWithGRPC creates a new Reader with the given range that uses
+// gRPC to read Object content.
+func (o *ObjectHandle) newRangeReaderWithGRPC(ctx context.Context, offset, length int64) (r *Reader, err error) {
+	ctx = trace.StartSpan(ctx, "cloud.google.com/go/storage.Object.newRangeReaderWithGRPC")
+	defer func() { trace.EndSpan(ctx, err) }()
+
+	if o.c.gc == nil {
+		err = fmt.Errorf("handle doesn't have a gRPC client initialized")
+		return
+	}
+	if err = o.validate(); err != nil {
+		return
+	}
+	if length < 0 {
+		err = fmt.Errorf("length must be non-negative")
+		return
+	}
+
+	req := &storagepb.ReadObjectRequest{
+		Bucket: o.bucket,
+		Object: o.object,
+	}
+
+	// Define a function that initiates a Read with offset and length, assuming we
+	// have already read seen bytes.
+	reopen := func(seen int64) (storagepb.Storage_ReadObjectClient, context.CancelFunc, error) {
+		// If the context has already expired, return immediately without making a
+		// call.
+		if err := ctx.Err(); err != nil {
+			return nil, nil, err
+		}
+
+		cc, cancel := context.WithCancel(ctx)
```

Comment:
Does this continue to respect any cancel that was handed out to users? (i assume yes...)

Reply:
If i understand your question, then yes. If the user has a `CancelFunc` for the context they gave the reader, it will still work as expected. And inversely, canceling this derived context will not cancel the user provided context. LMK if I misunderstood.

Reply:
Yep that's what I meant! Just thought I'd double check.